streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java

```diff
@@ -135,15 +133,19 @@ int addRawRecords(final TopicPartition partition, final Iterable<ConsumerRecord<
      * partition timestamp among all its partitions
      */
     public long timestamp() {
+        return streamTime;
+    }
+
+    private void computeStreamTime() {
         // we should always return the smallest timestamp of all partitions
         // to avoid group partition time goes backward
         long timestamp = Long.MAX_VALUE;
         for (final RecordQueue queue : partitionQueues.values()) {
-            if (timestamp > queue.timestamp()) {
+            if (queue.timestamp() < timestamp) {
                 timestamp = queue.timestamp();
             }
         }
-        return timestamp;
+        this.streamTime = timestamp;
```

Comment:
We could further optimize this computation by maintaining the queues in sorted order by their timestamp. Then, we would only ever have to look at the first queue in the sorted collection to find the new stream time.

However, this requires more bookkeeping during put and poll, so I'd like to avoid it for now. We can make this optimization if this method is a bottleneck (but I doubt it will be).